test/extended/builds/forcepull.go

```diff
@@ -38,152 +54,356 @@ while this test is running and compare results.  Restarting your docker daemon,
+				hexIDs, ierr := exutil.GetImageIDForTags(tags)
+				o.Expect(ierr).NotTo(o.HaveOccurred())
+				for i, hexID := range hexIDs {
+					g.By(fmt.Sprintf("\n FORCE PULL TEST  tag %s hex id %s ", tags[i], hexID))
+				}
+
+				// corrupt the s2i builder image
+				exutil.CorruptImage(fullName, custBldr, "s21")
+
+				// we need to plug in the dynamic values of the internal docker registry ip/port and namespace into the docker image spec
+				fpFalseS2Ipre := exutil.FixturePath("fixtures", falseS2IBC)
+				fpFalseS2Ipost := exutil.ArtifactPath(falseS2IBC)
+				err = exutil.VarSubOnFile(fpFalseS2Ipre, fpFalseS2Ipost, falseS2IBldr, fullName)
+				o.Expect(err).NotTo(o.HaveOccurred())
+
+				// kick off the app/lang build and verify the builder image accordingly
+				exutil.StartBuild(fpFalseS2Ipost, buildPrefix, oc)
+				exutil.WaitForBuild("FORCE PULL TEST s2i false app/lang build 1:  ", buildName, oc)
+				exutil.VerifyImagesSame(fullName, custBldr, "FORCE PULL TEST s2i false:  ")
 
 			})
 		})
 
-		g.Context("\n FORCE PULL TEST:  when s2i force pull is true and the image is bad", func() {
+		g.Context("\n FORCE PULL TEST  when s2i force pull is true and the image is bad", func() {
+			var oc = exutil.NewCLI("force-pull-s2i-true", exutil.KubeConfigPath())
+
+			g.JustBeforeEach(func() {
+				g.By("waiting for builder service account")
+				err := exutil.WaitForBuilderAccount(oc.KubeREST().ServiceAccounts(oc.Namespace()))
+				o.Expect(err).NotTo(o.HaveOccurred())
+			})
+
 			g.It("\n FORCE PULL TEST s2i true", func() {
-				fpTrueS2I := exutil.FixturePath("fixtures", "forcepull-true-s2i.json")
-				g.By(fmt.Sprintf("\n%s FORCE PULL TEST s2i true:  calling create on %s", time.Now().Format(time.RFC850), fpTrueS2I))
-				exutil.StartBuild(fpTrueS2I, buildPrefix, oc)
 
-				exutil.WaitForBuild("FORCE PULL TEST s2i true: ", buildName, oc)
+				// create the image streams
+				isTrueS2I := exutil.FixturePath("fixtures", trueS2IIS)
+				exutil.CreateResource(isTrueS2I, oc)
+
+				// kick off the build for the new builder image we can corrupt in parallel without conflicting with
+				// any other tests potentially running in parallel
+				bldrTrueS2I := exutil.FixturePath("fixtures", trueS2IBldrPrefix+".json")
+				exutil.StartBuild(bldrTrueS2I, trueS2IBldrPrefix, oc)
+				exutil.WaitForBuild("FORCE PULL TEST  s2i true bldr build:  ", trueS2IBldrPrefix+"-1", oc)
+
+				// get the auth so we can pull the build image from the internal docker registry since docker builder image will  removed it
+				// from the docker daemon cache when the docker build complete; we'll also get the internal docker registry's IP and port
+				// for pulling and tagging
+				authCfg, err := exutil.ConstructAuthCfg(oc)
+
+				// now actually pull the image back in from the openshift internal docker registry
+				fullName := authCfg.ServerAddress + "/" + oc.Namespace() + "/" + trueS2IBldr
+				err = exutil.PullImage(fullName, *authCfg)
+				o.Expect(err).NotTo(o.HaveOccurred())
```

Comment:
The shift to using the same builder image for multiple builds is not going well.  After the first s2i build (forcepull false) goes fine, the s2i build (forcepull true) does not work the way we want.  Here is a sequence of events from various logs. 

1) here is some debug from the ginkgo test after the bldr builds are constructed, before any corruption, noting their id's:

`[1] STEP: 
[1]  FORCE PULL TEST  tag 172.30.229.186:5000/extended-test-forcepull-mr0hp/dock-s2i-forcepull-ruby-20-centos7:latest hex id a01505bae1e05fc16755669071fbf3977a82a7b6247a2b9bdce5fb89283e5a00 
[1] STEP: 
[1]  FORCE PULL TEST  tag 172.30.229.186:5000/extended-test-forcepull-mr0hp/cust-forcepull-origin-custom-docker-builder:latest hex id 3692e0e81efc280f30d0243c2e858a2de5a9cf8563cbee1095e3d1e049b28276 
[1] STEP: 
[1]  FORCE PULL TEST  tag docker.io/openshift/origin-custom-docker-builder:latest hex id e2f857da8e3355f05156aaa368fd4c4b8de5b0644e99ff3169a2ed45a8d02a52 
`

2) Then here is more debug from the ginkgo test when those images are reset, to repair the corruption done by the s2i false test:

`[1] STEP: 
[1] Thursday, 24-Sep-15 19:33:57 EDT  Reset for 172.30.229.186:5000/extended-test-forcepull-mr0hp/dock-s2i-forcepull-ruby-20-centos7 to a01505bae1e05fc16755669071fbf3977a82a7b6247a2b9bdce5fb89283e5a00 complete with err <nil>
[1] STEP: 
[1] Thursday, 24-Sep-15 19:33:57 EDT  Reset for 172.30.229.186:5000/extended-test-forcepull-mr0hp/cust-forcepull-origin-custom-docker-builder to 3692e0e81efc280f30d0243c2e858a2de5a9cf8563cbee1095e3d1e049b28276 complete with err <nil>
[1] STEP: 
[1] Thursday, 24-Sep-15 19:33:57 EDT  Reset for docker.io/openshift/origin-custom-docker-builder to e2f857da8e3355f05156aaa368fd4c4b8de5b0644e99ff3169a2ed45a8d02a52 complete with err <nil>
`
3) Then, we corrupt again, and start the s2i true build.  Here are some snippets from the build controller pod's log.  I'll include the snippet that shows build config, and then the attempt to pull the builder image:

`I0924 23:34:42.891327       1 builder.go:37] $BUILD env var is {"kind":"Build","apiVersion":"v1","metadata":{"name":"ruby-sample-build-td-1","namespace":"extended-test-forcepull-9rhc0","selfLink":"/oapi/v1/namespaces/extended-test-forcepull-9rhc0/builds/ruby-sample-build-td-1","uid":"d46c0ee9-6314-11e5-a0c3-68f7287398a7","resourceVersion":"641","creationTimestamp":"2015-09-24T23:34:41Z","labels":{"buildconfig":"ruby-sample-build-td","name":"ruby-sample-build-td"},"annotations":{"openshift.io/build.number":"1"}},"spec":{"serviceAccount":"builder","source":{"type":"Git","git":{"uri":"https://github.com/openshift/ruby-hello-world.git"}},"strategy":{"type":"Docker","dockerStrategy":{"from":{"kind":"DockerImage","name":"172.30.229.186:5000/extended-test-forcepull-mr0hp/dock-s2i-forcepull-ruby-20-centos7"},"pullSecret":{"name":"builder-dockercfg-e3no5"},"forcePull":true}},"output":{},"resources":{}},"status":{"phase":"Pending","config":{"kind":"BuildConfig","namespace":"extended-test-forcepull-9rhc0","name":"ruby-sample-build-td"}}} 
`
`Step 0 : FROM 172.30.229.186:5000/extended-test-forcepull-mr0hp/dock-s2i-forcepull-ruby-20-centos7
Trying to pull repository 172.30.229.186:5000/extended-test-forcepull-mr0hp/dock-s2i-forcepull-ruby-20-centos7 ... not found
F0924 23:34:44.097580       1 builder.go:54] Build error: Error: image extended-test-forcepull-mr0hp/dock-s2i-forcepull-ruby-20-centos7:latest not found
Unable to find user default
`
I see a similar error when I try to manually pull the image from the ginkgo code after the s2i true test runs.

While this is occurring though, I can see the force pull builder image when I run docker images from the cmd line (with the right hex id).
 
Close to crying uncle here ..... any suggestions? .... thanks